ksqldb-engine/src/main/java/io/confluent/ksql/internal/StorageUtilizationMetricsReporter.java

```diff
@@ -0,0 +1,288 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.metrics.MetricCollectors;
+import java.io.File;
+import java.math.BigInteger;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import io.confluent.ksql.util.KsqlException;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.MetricsContext;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.streams.StreamsConfig;
+
+public class StorageUtilizationMetricsReporter implements MetricsReporter {
+  private static final String METRIC_GROUP = "ksqldb_utilization";
+
+  private final Map<String, Map<String, TaskStorageMetric>> metricsSeen;
+  private final Metrics metricRegistry;
+  static AtomicBoolean registeredNodeMetrics = new AtomicBoolean(false);
+
+  public StorageUtilizationMetricsReporter() {
+    this(MetricCollectors.getMetrics());
+  }
+
+  public StorageUtilizationMetricsReporter(final Metrics metricRegistry) {
+    this.metricsSeen = new HashMap<>();
+    this.metricRegistry = metricRegistry;
+  }
+
+  @Override
+  public void init(final List<KafkaMetric> list) {
+  }
+
+  @Override
+  public void configure(final Map<String, ?> map) {
+    final String dir;
+    if (map.containsKey(StreamsConfig.STATE_DIR_CONFIG)) {
+      dir = map.get(StreamsConfig.STATE_DIR_CONFIG).toString();
+    } else {
+      dir =  StreamsConfig
+        .configDef()
+        .defaultValues()
+        .get(StreamsConfig.STATE_DIR_CONFIG)
+        .toString();
+    }
+    configureShared(new File(dir), this.metricRegistry);
+  }
+
+  private static void configureShared(final File baseDir, final Metrics metricRegistry) {
+    if (registeredNodeMetrics.getAndSet(true)) {
+      return;
+    }
+    final MetricName nodeAvailable =
+        metricRegistry.metricName("node_storage_available_bytes", METRIC_GROUP);
+    final MetricName nodeTotal =
+        metricRegistry.metricName("node_storage_total_bytes", METRIC_GROUP);
+    final MetricName nodeUsed =
+        metricRegistry.metricName("node_storage_used_bytes", METRIC_GROUP);
+
+    metricRegistry.addMetric(
+        nodeAvailable,
+        (Gauge<Long>) (config, now) -> baseDir.getFreeSpace()
+    );
+    metricRegistry.addMetric(
+        nodeTotal,
+        (Gauge<Long>) (config, now) -> baseDir.getTotalSpace()
+    );
+    metricRegistry.addMetric(
```

Comment:
can you add a `storage_utilization` metric that returns ((baseDir.getTotalSpace - baseDir.getFreeSpace) / baseDir.getTotalSpace)

Reply:
There should be no risk of division by 0 here, right? Since our `baseDir` should always exist

Reply:
And do you want this all the way up to the metrics api + exposed to users?

Reply:
> There should be no risk of division by 0 here, right? Since our baseDir should always exist

yeah that should never happen

> And do you want this all the way up to the metrics api + exposed to users?

yeah, the max value of this metric (over all the nodes) is what we want to show users in cloud